istioctl/cmd/sidecar.go

```diff
@@ -0,0 +1,291 @@
+				if err != nil {
+					return fmt.Errorf("could not infer (project, location, cluster). Try passing in flags instead")
+				}
+			}
+
+			if err = createConfig(wg, cluster, outputName); err != nil {
+				return err
+			}
+			fmt.Printf("generated config for (%s, %s, %s) in %s\n", gkeProject, clusterLocation, clusterName, outputName)
+			return nil
+		},
+	}
+	generateConfigCmd.PersistentFlags().StringVarP(&filename, "file", "f", "", "filename of the WorkloadGroup artifact")
+	generateConfigCmd.PersistentFlags().StringVarP(&outputName, "output", "o", "", "Name of the tarball to be created")
+
+	generateConfigCmd.PersistentFlags().StringVar(&gkeProject, "project", "", "Target project name")
+	generateConfigCmd.PersistentFlags().StringVar(&clusterLocation, "location", "", "Target cluster location")
+	generateConfigCmd.PersistentFlags().StringVar(&clusterName, "cluster", "", "Target cluster name")
+	return generateConfigCmd
+}
+
+func readWorkloadGroup(filename string, wg *clientv1alpha3.WorkloadGroup) error {
+	f, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return err
+	}
+	if err = yaml.Unmarshal(f, wg); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Creates all the relevant config for the given workload group and cluster
+func createConfig(wg *clientv1alpha3.WorkloadGroup, cluster *containerpb.Cluster, outputName string) error {
+	temp, err := ioutil.TempDir("./", outputName)
+	if err != nil {
+		return err
+	}
+	// TODO: pack temp into a tarball and then delete folder
+	// defer os.RemoveAll(temp)
+
+	if err = createClusterEnv(wg, cluster, temp); err != nil {
+		return err
+	}
+	if err = createHosts(temp); err != nil {
+		return err
+	}
+	if err = createCertificates(temp); err != nil {
+		return err
+	}
+	if err = createMeshConfig(wg, temp); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Write cluster.env into the given directory
+func createClusterEnv(wg *clientv1alpha3.WorkloadGroup, cluster *containerpb.Cluster, dir string) error {
+	for _, v := range wg.Spec.Ports {
+		ports = append(ports, fmt.Sprint(v))
+	}
+	// default attributes and service name, namespace, ports, service account, service CIDR
+	clusterEnv := map[string]string{
+		"ISTIO_CP_AUTH":       "MUTUAL_TLS",
+		"ISTIO_INBOUND_PORTS": strings.Join(ports, ","),
+		"ISTIO_NAMESPACE":     wg.Namespace,
+		"ISTIO_PILOT_PORT":    "15012",
+		"ISTIO_SERVICE":       fmt.Sprintf("%s.%s", wg.Name, wg.Namespace),
+		"ISTIO_SERVICE_CIDR":  cluster.ServicesIpv4Cidr,
+		"SERVICE_ACCOUNT":     wg.Spec.ServiceAccount,
+	}
+	return ioutil.WriteFile(dir+"/cluster.env", []byte(mapToString(clusterEnv)), tempPerms)
+}
+
+// Create the needed hosts addition in the given directory
+func createHosts(dir string) error {
+	istiod, err := k8sService("istiod", "istio-system")
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(dir+"/hosts", []byte(fmt.Sprintf("%s\t%s", istiod.Spec.ClusterIP, "istiod.istio-system.svc")), tempPerms)
```

Comment:
ClusterIP is not reachable from a VM, so I don't think this will work

Reply:
It seems like in the VM setup guide, we add IstiodIP to the hosts file.
`echo "${IstiodIP} istiod.istio-system.svc" | sudo tee -a /etc/hosts`

Reply:
Ack, see my comment about how you have faithfully reproduce the docs, but the docs are wrong :slightly_smiling_face:

Reply:
ya - that part about the `sudo tee -a /etc/hosts` is actually pretty bad - and on that point we agree @howardjohn - this is where the docs fall over.

I'll see if I can sort out a better DNS procedure early next week that does not require code change. I have some good ideas for how to handle DNS properly - but I am uncertain if code change is required or not.